docs/order-confirmation/index.md

```diff
@@ -0,0 +1,333 @@
+
+- Add a `Hotdog` constructor to `Order.t`
+- Add a `| Hotdog` branch to the switch expression of `Order.toPrice`
+
+Your OCaml code wouldn't compile if you just added a new constructor to
+`Item.t` without also updating `Item.toPrice`. Adding or removing constructors
+from a variant usually forces you to change relevant parts of your code.
+
+## `fun` sugar syntax
+
+Incidentally, there's a sugar syntax for functions whose entire body is a switch
+expression. It's called `fun`, and we can rewrite `Item.toPrice` to use it:
+
+<<< Snippets.re#to-price-fun
+
+We can also define `toEmoji` using the `fun` sugar syntax:
+
+<<< Snippets.re#to-emoji
+
+Using the `fun` syntax is completely equivalent to using a switch expression,
+so it's up to your personal taste whether you want to use this sugar syntax.
+
+## `Order.make`
+
+Now we're ready to define the `make` function which will render the `Item`
+component:
+
+<<< Snippets.re#make
+
+The `make` function has a single labeled argument, `~item`, of type `t`. This
+effectively means the `Item` component has a single prop named `item`.
+
+Note that this renders a single row of a table. We'll need another component
+to render a table containing all items in an order.
+
+## `Order` component
+
+Create a new file `src/order-confirmation/Order.re` and add the following code:
+
+<<< Snippets.re#order
+
+There's a lot going on here:
+
+- The primary type of `Order` is `array(Item.t)`, which is an array of
+  variants.
+- The `make` function has a single labeled argument, `~items`, of type `t`.
+  This means the `Order` component has a single prop named `items`.
+- We sum up the prices of all items using
+  [Js.Array.reduce](https://melange.re/v2.1.0/api/re/melange/Js/Array/index.html#val-reduce).
+  This is identical to JavaScript's [Array.reduce
+  method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce),
+  except that the Melange version requires the initial value to be passed in.
+- For each order, we render an `Item` component via
+  [Js.Array.map](https://melange.re/v2.1.0/api/re/melange/Js/Array/index.html#val-map),
+  which is equivalent to the [Array.map
+  method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).
+- The `React.array` function has the type signature `array(React.element) =>
+  React.element`, meaning it converts an array of `React.element`s to
+  `React.element`. All expressions in JSX must have the type `React.element`.
+
+## `Index.re`
+
+Render the `Order` component inside `Index.re`:
+
+<<< Snippets.re#index
+
+Run `make serve` inside `src/order-confirmation` to see your new app in action.
+Open your browser's dev console, where you should see a warning:
+
+```
+Warning: Each child in a list should have a unique "key" prop.
+```
+
+Oops, we should've used
+[Js.Array.mapi](https://melange.re/v2.1.0/api/re/melange/Js/Array/index.html#val-mapi)[^3]
```

Comment:
Not sure if we should do this, in general it's not recommended to use indexes as keys, or generate the keys at render time: https://react.dev/learn/rendering-lists#rules-of-keys

Maybe the order could be a record `{id: 1, item: Sandwich}`?

Reply:
This is a very good point, but I think it's a "useful mistake". I think I should stick to my plan of introducing records in the next chapter, and fix it there. Wdyt?

Reply:
Sounds good. Maybe briefly mention something like "this is not recommended per [this link](https://react.dev/learn/rendering-lists#rules-of-keys) but we'll fix it soon" might give readers a heads up?

Reply:
I added a footnote about this in https://github.com/melange-re/melange-for-react-devs/pull/9/commits/8e4c39fbd9849a3e57d9b01841fc85190b6d3adb

```
[^4]: Using array indexes to set keys violates [React's rules of
    keys](https://react.dev/learn/rendering-lists#rules-of-keys), which states
    that you shouldn't generate keys while rendering. We'll see a better way to
    do this [later](/todo).
```